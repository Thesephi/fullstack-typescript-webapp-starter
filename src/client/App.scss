@import "~@fontsource/roboto/index.css";

.the-dialog {
  .MuiDialog-paper {
    max-width: 660px;
  }
  h5 {
    margin: 0;
  }
}
