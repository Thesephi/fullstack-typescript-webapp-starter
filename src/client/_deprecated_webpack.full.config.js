const webpack = require("webpack");
const path = require("path");
const HtmlWebpackPlugin = require("html-webpack-plugin");

const isDevelopment = process.env.NODE_ENV !== "production";
const mode = isDevelopment ? "development" : "production";

console.log(`webpack operating under mode \`${mode}\``);

module.exports = {

    mode,
    watch: process.env.WATCH === "true",

    target: "web",
    entry: {
        main: [
        "webpack-hot-middleware/client?path=__webpack_hmr&timeout=2000&overlay=false",
        path.resolve(__dirname, "./main.tsx")
        // "src/client/main.tsx"
        ]
    },

    output: {
        // path: path.resolve(__dirname, "../../build/public"),
        // filename: "js/main.js"
        path: process.env.WEBPACK_CLIENT_APP_OUTPUT_DIR,
        filename: process.env.WEBPACK_CLIENT_APP_OUTPUT_FILENAME,
        publicPath: "/"
    },

    devServer: {
        contentBase: path.resolve(__dirname, "../../build/public"),
        hot: true
    },

    // Enable sourcemaps for debugging webpack's output.
    devtool: "source-map",

    resolve: {
        // Add '.ts' and '.tsx' as resolvable extensions.
        // Add '.js' is so that React itself can be compiled (though this is not required during prod).
        extensions: [".ts", ".tsx", ".js"],
        alias: {
        shared: path.resolve(__dirname, "shared/")
        }
    },

    module: {
        rules: [
            {
                // test: /\.ts(x?)$/,
                test: /\.(t|j)s(x?)$/,
                // include: path.resolve(process.cwd(), "src/client"),
                exclude: /node_modules/,
                use: [
                    isDevelopment && {
                    loader: "babel-loader",
                    options: { plugins: ["react-refresh/babel"] }
                    },
                    {
                    loader: "ts-loader",
                    options: { transpileOnly: true }
                    }
                ].filter(Boolean)
            },
            {
                test: /\.s[ac]ss$/i,
                use: [
                // Creates `style` nodes from JS strings
                "style-loader",
                // Translates CSS into CommonJS
                "css-loader",
                // Compiles Sass to CSS
                "sass-loader"
                ]
            },
            {
                test: /\.(woff|woff2|eot|ttf|otf|node)$/,
                loader: "file-loader",
                options: {
                outputPath: "fonts"
                }
            },
            // All output '.js' files will have any sourcemaps re-processed by 'source-map-loader'.
            {
                enforce: "pre",
                test: /\.js$/,
                loader: "source-map-loader"
            }
        ]
    },

    plugins: [
        new webpack.EnvironmentPlugin({
        NODE_ENV: mode,
        APP_NAME: process.env.APP_NAME
        }),
        new HtmlWebpackPlugin({
            filename: path.resolve(__dirname, "../../build/server/view-templates/index.html"),
            template: path.resolve(__dirname, "../server/view-templates/index.html"),
            publicPath: "/"
        }),
        isDevelopment && new webpack.HotModuleReplacementPlugin(), // disable if using webpack-dev-server
        isDevelopment && new (require("@pmmmwh/react-refresh-webpack-plugin"))({
            overlay: {
                sockIntegration: "whm",
            }
        })
    ]

};
