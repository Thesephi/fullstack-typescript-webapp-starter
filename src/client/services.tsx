import { IRespPayload } from "shared/types";

interface IEntry {
    firstName: string;
    lastName: string;
    email: string;
    // eventDate: Date;
}

export function submitEntry(entry: IEntry): Promise<IRespPayload> {
    return new Promise((resolve, reject) => {

        window.fetch("/entry", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                firstName: entry.firstName.trim(),
                lastName: entry.lastName.trim(),
                email: entry.email.trim(),
                // eventDate: entry.eventDate.getTime()
            })
        })
        .then(resp => resolve(resp.json()))
        .catch(reject);

    });
}
