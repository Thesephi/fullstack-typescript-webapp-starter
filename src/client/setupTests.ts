// see https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html
(globalThis as any).IS_REACT_ACT_ENVIRONMENT = true
