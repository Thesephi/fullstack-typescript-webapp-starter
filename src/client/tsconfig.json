{
    "compilerOptions": {
        "outDir": "../../build/public/",
        "baseUrl": ".",
        "paths": {
          "shared/*": ["../shared/*"]
        },
        "sourceMap": true,
        "esModuleInterop": true,
        "noImplicitAny": true,
        "module": "commonjs",
        "target": "es6",
        "jsx": "react",
        "strict": true,
        "lib": [
            "dom",
            "dom.iterable",
            "esnext"
        ]
    }
}
