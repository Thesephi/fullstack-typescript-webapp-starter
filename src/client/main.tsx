import * as React from "react";
import { Root, createRoot } from "react-dom/client";
import App from "./App";

const root: Root = createRoot(document.getElementById("main")!);
root.render(<App />);

declare var module: {
    hot?: any;
}

if (module.hot) {
    module.hot.accept("./App", (err: unknown) => {
        try {
            const NextApp = require("./App").default;
            root.render(<NextApp />);
            return;
        } catch (e) {
            console.error(e);
            // to be reloaded
        }
        console.error(err);
        console.log("reloading self");
        window.location.reload();
    });
}