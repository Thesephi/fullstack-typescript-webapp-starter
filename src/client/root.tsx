/**
 * This should be loaded into an HTML page that was server-rendered by React (via `withRootApp`)
 */

import React from "react";
import { hydrateRoot } from "react-dom/client";
import { RootApp } from "shared/RootApp";
// import { ExampleReactComponent } from "shared/domains/ExampleReactComponent";

declare var window: Window & {
  __PRELOADED_STATE__: any;
}

const preloadedState = window.__PRELOADED_STATE__;

hydrateRoot(document, <RootApp preloadedState={preloadedState} />);