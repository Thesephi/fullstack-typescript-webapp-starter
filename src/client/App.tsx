import * as React from "react";
import { useState } from "react";
import Button from "@mui/material/Button";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import DialogContentText from "@mui/material/DialogContentText";
import DialogTitle from "@mui/material/DialogTitle";
import Input from "@mui/material/Input";
import { Alert, AlertTitle } from "@mui/material";
import "./App.scss";

import { isValidEmail } from "shared/utils";
import { submitEntry } from "./services";
import { IRespPayload } from "shared/types";

const App: React.FunctionComponent = () => {

  const [firstName, setFirstName] = useState<string>("");
  const [lastName, setLastName] = useState<string>("");
  const [email, setEmail] = useState<string>("");
  // const [eventDate, setEventDate] = useState<Date>(new Date());
  const [errors, setErrors] = useState<string[]>([]);
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
  const [feedback, setFeedback] = useState<string>("");

  function validateAllFields(): true | string[] {
      const validationIssues: string[] = [];
      if (firstName === "") validationIssues.push("missing first name");
      if (lastName === "") validationIssues.push("missing last name");
      if (!isValidEmail(email)) validationIssues.push("invalid email address");
      // if (eventDate == null) validationIssues.push("missing event date");
      return validationIssues.length ? validationIssues : true;
  }

  async function attemptSubmit(evt?: React.FormEvent<HTMLFormElement>): Promise<void> {

      if (isSubmitting) return;

      if (evt && evt.type === "submit") {
        // we handle form submission instead of letting the browser redirect
        // as per its default behavior
        evt.preventDefault();
      }

      // clear error and feedback views before making a new submission
      setErrors([]);
      setFeedback("");

      const validationResult: true | string[] = validateAllFields();
      if (validationResult === true) {

          setIsSubmitting(true);
          try {

              // send all fields to backend
              const resp: IRespPayload = await submitEntry({ firstName, lastName, email });
              setFeedback(resp.message);
              
          } catch (e) {
              // display API call error
              setErrors([`An error occurred: ${(e as Error).message}`]);
          }
          setIsSubmitting(false);

      } else {
          // display validation error(s)
          setErrors(validationResult);
      }

  }

  return <Dialog
    open={true}
    className="the-dialog"
    aria-labelledby="dialog-title"
    aria-describedby="dialog-description"
  >
  <form action="#" autoComplete="off" onSubmit={attemptSubmit}>
    <DialogTitle id="dialog-title">{process.env.APP_NAME}</DialogTitle>
    <DialogContent>
      <h5><DialogContentText>
        <small>{"all fields are required"}</small>
      </DialogContentText></h5>
        <div style={{ display: "flex", flexDirection: "row" }}>
          <div style={{ flexGrow: 1 }}>
            <Input
              name="firstName"
              type="text"
              inputMode="text"
              placeholder={"first name"}
              value={firstName}
              onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setFirstName(evt.target.value)}
              inputProps={{ "aria-label": "first name" }}
              fullWidth
            />
          </div>
          <div style={{ minWidth: 10, flexGrow: 0 }} />
          <div style={{ flexGrow: 1 }}>
            <Input
              name="lastName"
              type="text"
              inputMode="text"
              placeholder={"last name"}
              value={lastName}
              onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setLastName(evt.target.value)}
              inputProps={{ "aria-label": "last name" }}
              fullWidth
            />
          </div>
        </div>
        <div style={{ marginTop: 10, marginBottom: 10 }}>
          <Input
            name="email"
            type="email"
            inputMode="email"
            placeholder={"email"}
            value={email}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setEmail(evt.target.value)}
            inputProps={{ "aria-label": "email" }}
            fullWidth
          />
        </div>
      {feedback !== "" && <Alert severity="info" className="feedback-info">
        <span>{feedback}</span>
        {" "}
        (<a href="/entries" target="_blank" style={{ textDecoration: "underline", color: "inherit" }}>
            {"view all entries as JSON"}
        </a>)
      </Alert>}
      {errors.length > 0 && <Alert severity="error" className="feedback-error">
        <AlertTitle>{"You have entered invalid information"}</AlertTitle>
        <ul style={{ padding: "0 0 0 20px", margin: 0 }}>
            {errors.map((error, i) => <li key={i}>{error}</li>)}
        </ul>
      </Alert>}
    </DialogContent>
    <DialogActions>
      <Button type="submit" color="primary" autoFocus disableRipple>
        {"Submit"}
      </Button>
    </DialogActions>
  </form>
  </Dialog>;

};

export default App;

// below are alternative (examplary) approaches to rendering (almost) the same
// layout using different ReactJS UI frameworks:

// 1. Blueprint
/*
return (
<Dialog
  isOpen={true}
  title={process.env.APP_NAME}
  isCloseButtonShown={false}
  style={{ minWidth: 600 }}
>
  <div className={Classes.DIALOG_BODY}>
    <H5>Demo simple web app <small>(all fields are required)</small></H5>
    <form action="#">
      <div style={{ display: "flex", flexDirection: "row" }}>
        <div style={{ flexGrow: 1 }}>
          <InputGroup
            name="firstName"
            type="text"
            inputMode="text"
            fill
            placeholder={"first name"}
            value={firstName}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setFirstName(evt.target.value)}
          />
        </div>
        <div style={{ minWidth: 10, flexGrow: 0 }} />
        <div style={{ flexGrow: 1 }}>
          <InputGroup
            name="lastName"
            type="text"
            inputMode="text"
            fill
            placeholder={"last name"}
            value={lastName}
            onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setLastName(evt.target.value)}
          />
        </div>
      </div>
      <InputGroup
          style={{ marginTop: 10, marginBottom: 10 }}
          name="email"
          type="email"
          inputMode="email"
          placeholder={"email"}
          value={email}
          onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setEmail(evt.target.value)}
      />
    </form>
    {feedback !== "" && <Callout style={{ marginTop: 10 }} className="co-feedback">
        <span>{feedback}</span>
        {" "}
        (<a href="/entries" target="_blank" style={{ textDecoration: "underline", color: "inherit" }}>
            view all entries as JSON
        </a>)
    </Callout>}
    {errors.length > 0 && <Callout intent={Intent.DANGER} icon={null} style={{ marginTop: 10 }} className="co-errors">
        <ul style={{ padding: "0 0 0 20px", margin: 0 }}>
            {errors.map((error, i) => <li key={i}>{error}</li>)}
        </ul>
    </Callout>}
  </div>
  <div className={Classes.DIALOG_FOOTER} style={{ display: "flex" }}>
      <span style={{ opacity: 0.5 }}>client env: {process.env.NODE_ENV}</span>
      <span style={{ flexGrow: 1 }} />
      <Button type="submit" text={"Submit"} intent={Intent.PRIMARY} onClick={attemptSubmit} autoFocus />
  </div>
</Dialog>
)
*/

// 2. Atlaskit
/*
const CustomFooter = (props: FooterComponentProps) => {
  return (
    <ModalFooter {...props}>
      <span />
      <Button appearance={props.appearance} onClick={attemptSubmit}>
        Submit
      </Button>
    </ModalFooter>
  );
};

return <div><ModalTransition>
  <Modal
    components={{
      Footer: CustomFooter,
      Container: ({ children }) => <div className="model-container">
        <div className="model-host">{children}</div>
      </div>
    }}
    heading={process.env.APP_NAME}
  >
    <h5>Demo simple web app <small>(all fields are required)</small></h5>
    <form action="#">
    <div style={{ display: "flex", flexDirection: "row" }}>
      <div style={{ flexGrow: 1 }}>
        <Textfield
          name="firstName"
          type="text"
          inputMode="text"
          placeholder={"first name"}
          value={firstName}
          onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setFirstName(evt.target.value)}
          isCompact
        />
      </div>
      <div style={{ minWidth: 10, flexGrow: 0 }} />
      <div style={{ flexGrow: 1 }}>
        <Textfield
          name="lastName"
          type="text"
          inputMode="text"
          placeholder={"last name"}
          value={lastName}
          onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setLastName(evt.target.value)}
          isCompact
        />
      </div>
    </div>
    <div style={{ marginTop: 10, marginBottom: 10 }}>
      <Textfield
        name="email"
        type="email"
        inputMode="email"
        placeholder={"email"}
        value={email}
        onChange={(evt: React.ChangeEvent<HTMLInputElement>) => setEmail(evt.target.value)}
        isCompact
      />
    </div>
    </form>
    {feedback !== "" && <SectionMessage>
      <span>{feedback}</span>
      {" "}
      (<a href="/entries" target="_blank" style={{ textDecoration: "underline", color: "inherit" }}>
          view all entries as JSON
      </a>)
    </SectionMessage>}
    {errors.length > 0 && <SectionMessage appearance="error" title={"You have entered invalid information"}>
      <ul style={{ padding: "0 0 0 20px", margin: 0 }}>
          {errors.map((error, i) => <li key={i}>{error}</li>)}
      </ul>
    </SectionMessage>}
  </Modal>
</ModalTransition></div>;
*/
