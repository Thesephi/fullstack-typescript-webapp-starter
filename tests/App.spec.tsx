/**
 * @jest-environment jsdom
 */

import "jest";
import * as React from "react";
import { act, Simulate } from "react-dom/test-utils";
import { createRoot, Root } from "react-dom/client";
import App from "../src/client/App";

describe("App", () => {

  let root: Root;
  let container: HTMLDivElement | null;

  let defaultFetch: any = window.fetch;
  let mockFetch = jest.fn((entry) => Promise.resolve({
    json: () => Promise.resolve({ message: `Yay`, details: { foo: "bar" } })
  }));

  beforeAll(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    root = createRoot(container!);
    act(() => {
      root.render(<App />);
    });
    window.fetch = mockFetch as any;
  });

  afterAll(() => {
    act(() => {
      root.unmount();
    });
    document.body.removeChild(container!);
    container = null;
    window.fetch = defaultFetch;
  });

  it("should render the dialog DOM as specified", () => {
    const dialog = document.querySelector(".the-dialog")!;
    expect(dialog.innerHTML).toMatchSnapshot();
  });

  it("should not show any error in the beginning", () => {
    const calloutErrors = document.querySelector(".feedback-error");
    expect(calloutErrors).toBeNull();
  });

  it("should show errors if the submit button is clicked when all fields are still invalid", () => {

    const submitBtn = document.querySelector("button[type=submit]") as HTMLButtonElement;
    expect(submitBtn).not.toBeNull();

    const form: HTMLFormElement = document.querySelector("form") as HTMLFormElement;
    expect(form.contains(submitBtn)).toEqual(true);

    // @TODO find out why simulating the `click` event on the submit button does
    // not work, but simulating the `submit` event on the form does
    // const submitBtn = document.querySelector("button[type=submit]") as HTMLButtonElement;
    // await act(async () => {
    //   Simulate.click(submitBtn);
    // });

    act(() => {
      Simulate.submit(form);
    });

    expect(mockFetch).not.toHaveBeenCalled();

    const calloutErrors = document.querySelector(".feedback-error") as HTMLElement;
    expect(calloutErrors).not.toBeNull();

    expect(calloutErrors.innerHTML).toMatchSnapshot();

  });

  it("should still show errors if the submit button is clicked when only 1 field is valid", async () => {

    const firstNameInput: HTMLInputElement = document.querySelector("input[name=firstName]") as HTMLInputElement;
    expect(firstNameInput).not.toBeNull();

    await act(async () => {
      firstNameInput.value = "John "; // the space is deliberate
      Simulate.change(firstNameInput);
    });

    const submitBtn = document.querySelector("button[type=submit]") as HTMLButtonElement;
    await act(async () => {
      Simulate.click(submitBtn);
    });

    expect(mockFetch).not.toHaveBeenCalled();

    const calloutErrors = document.querySelector(".feedback-error") as HTMLElement;
    expect(calloutErrors.innerHTML).toMatchSnapshot();

  });

  it("should still show errors if the submit button is clicked when only 2 fields are valid", async () => {

    const lastNameInput: HTMLInputElement = document.querySelector("input[name=lastName]") as HTMLInputElement;
    expect(lastNameInput).not.toBeNull();

    await act(async () => {
      lastNameInput.value = " Doe"; // the space is deliberate
      Simulate.change(lastNameInput);
    });

    const submitBtn = document.querySelector("button[type=submit]") as HTMLButtonElement;
    await act(async () => {
      Simulate.click(submitBtn);
    });

    expect(mockFetch).not.toHaveBeenCalled();

    const calloutErrors = document.querySelector(".feedback-error") as HTMLElement;
    expect(calloutErrors.innerHTML).toMatchSnapshot();

  });

  it("should make the POST call & show no errors if the submit button is clicked when all fields are valid", async () => {

    const emailInput: HTMLInputElement = document.querySelector("input[name=email]") as HTMLInputElement;
    expect(emailInput).not.toBeNull();

    /*
    const dateInput: HTMLInputElement = document.querySelector("input[name=eventDate]") as HTMLInputElement;
    expect(dateInput).not.toBeNull();

    // simulate a date being selected via the DatePicker
    const selectedDate: string = (new Date(dateInput.value)).toLocaleDateString();
    await act(async () => {
        dateInput.value = selectedDate;
        Simulate.change(dateInput);
    });
    */

    await act(async () => {
      emailInput.value = "john.doe@email.com";
      Simulate.change(emailInput);
    });

    // @TODO find out why simulating the `click` event on the submit button does
    // not work, but simulating the `submit` event on the form does
    // const submitBtn = document.querySelector("button[type=submit]") as HTMLButtonElement;
    // await act(async () => {
    //   Simulate.click(submitBtn);
    // });

    const form: HTMLFormElement = document.querySelector("form") as HTMLFormElement;
    await act(async () => {
      Simulate.submit(form);
    });

    expect(mockFetch).toHaveBeenCalledTimes(1);
    expect(mockFetch).toHaveBeenLastCalledWith("/entry", expect.objectContaining({
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        firstName: "John",
        lastName: "Doe",
        email: "john.doe@email.com",
        // eventDate: new Date(dateInput.value).getTime()
      })
    }));

    const calloutErrors = document.querySelector(".feedback-error") as HTMLElement;
    expect(calloutErrors).toBeNull();

    const calloutFeedback = document.querySelector(".feedback-info") as HTMLElement;
    expect(calloutFeedback.innerHTML).toMatchSnapshot();

  });

});
