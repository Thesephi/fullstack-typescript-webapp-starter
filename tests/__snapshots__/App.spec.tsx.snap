// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should make the POST call & show no errors if the submit button is clicked when all fields are valid 1`] = `"<div class=\\"MuiAlert-icon css-1ytlwq5-MuiAlert-icon\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root\\" focusable=\\"false\\" aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" data-testid=\\"InfoOutlinedIcon\\"><path d=\\"M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z\\"></path></svg></div><div class=\\"MuiAlert-message css-1pxa9xg-MuiAlert-message\\"><span>Yay</span> (<a href=\\"/entries\\" target=\\"_blank\\" style=\\"text-decoration: underline;\\">view all entries as JSON</a>)</div>"`;

exports[`App should render the dialog DOM as specified 1`] = `"<div aria-hidden=\\"true\\" class=\\"MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div><div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div><div class=\\"MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"presentation\\" tabindex=\\"-1\\"><div class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper\\" role=\\"dialog\\" aria-describedby=\\"dialog-description\\" aria-labelledby=\\"dialog-title\\"><form action=\\"#\\" autocomplete=\\"off\\"><h2 class=\\"MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root\\" id=\\"dialog-title\\">Test App</h2><div class=\\"MuiDialogContent-root css-ypiqx9-MuiDialogContent-root\\"><h5><p class=\\"MuiDialogContentText-root MuiTypography-root MuiTypography-body1 css-qfso29-MuiTypography-root-MuiDialogContentText-root\\"><small>all fields are required</small></p></h5><div style=\\"display: flex; flex-direction: row;\\"><div style=\\"flex-grow: 1;\\"><div inputmode=\\"text\\" class=\\"MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-sje21u-MuiInputBase-root-MuiInput-root\\"><input name=\\"firstName\\" placeholder=\\"first name\\" type=\\"text\\" aria-label=\\"first name\\" class=\\"MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input\\" value=\\"\\"></div></div><div style=\\"min-width: 10px; flex-grow: 0;\\"></div><div style=\\"flex-grow: 1;\\"><div inputmode=\\"text\\" class=\\"MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-sje21u-MuiInputBase-root-MuiInput-root\\"><input name=\\"lastName\\" placeholder=\\"last name\\" type=\\"text\\" aria-label=\\"last name\\" class=\\"MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input\\" value=\\"\\"></div></div></div><div style=\\"margin-top: 10px; margin-bottom: 10px;\\"><div inputmode=\\"email\\" class=\\"MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-sje21u-MuiInputBase-root-MuiInput-root\\"><input name=\\"email\\" placeholder=\\"email\\" type=\\"email\\" aria-label=\\"email\\" class=\\"MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root\\"><button class=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-focusVisible  css-1e6y48t-MuiButtonBase-root-MuiButton-root\\" tabindex=\\"0\\" type=\\"submit\\">Submit</button></div></form></div></div><div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div>"`;

exports[`App should show errors if the submit button is clicked when all fields are still invalid 1`] = `"<div class=\\"MuiAlert-icon css-1ytlwq5-MuiAlert-icon\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root\\" focusable=\\"false\\" aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" data-testid=\\"ErrorOutlineIcon\\"><path d=\\"M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"></path></svg></div><div class=\\"MuiAlert-message css-1pxa9xg-MuiAlert-message\\"><div class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root\\">You have entered invalid information</div><ul style=\\"padding: 0px 0px 0px 20px; margin: 0px;\\"><li>missing first name</li><li>missing last name</li><li>invalid email address</li></ul></div>"`;

exports[`App should still show errors if the submit button is clicked when only 1 field is valid 1`] = `"<div class=\\"MuiAlert-icon css-1ytlwq5-MuiAlert-icon\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root\\" focusable=\\"false\\" aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" data-testid=\\"ErrorOutlineIcon\\"><path d=\\"M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"></path></svg></div><div class=\\"MuiAlert-message css-1pxa9xg-MuiAlert-message\\"><div class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root\\">You have entered invalid information</div><ul style=\\"padding: 0px 0px 0px 20px; margin: 0px;\\"><li>missing first name</li><li>missing last name</li><li>invalid email address</li></ul></div>"`;

exports[`App should still show errors if the submit button is clicked when only 2 fields are valid 1`] = `"<div class=\\"MuiAlert-icon css-1ytlwq5-MuiAlert-icon\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root\\" focusable=\\"false\\" aria-hidden=\\"true\\" viewBox=\\"0 0 24 24\\" data-testid=\\"ErrorOutlineIcon\\"><path d=\\"M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"></path></svg></div><div class=\\"MuiAlert-message css-1pxa9xg-MuiAlert-message\\"><div class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root\\">You have entered invalid information</div><ul style=\\"padding: 0px 0px 0px 20px; margin: 0px;\\"><li>missing first name</li><li>missing last name</li><li>invalid email address</li></ul></div>"`;
